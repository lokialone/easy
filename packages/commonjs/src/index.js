// let r = require('./a');
// console.log(this) // => {}
// console.log(r);
// const a = 1;
// const b = 2;
// function sum(x,y){
//     return x+y;
// }
// sum(a,b)

// 1.掌握node中如何实现代码调试
// https://nodejs.org/en/docs/inspector
// 1.可以在浏览器中进行调试 （调试某些模块可以使用这种方式） node --inspect-brk 文件名来解析
// √ 2.直接使用webstorm 和 vscode 自带调试方式  直接通过launch.json 进行调试
// 3.在控制台中调试 

// 分析node源码
// 1.会默认调用require语法
// 2.Module.prototype.require 模块的原型上有require方法
// 3.Module._load 调用模块的加载方法  最终返回的是module.exports
// 4.Module._resolveFilename 解析文件名 将文件名变成绝对路径 默认尝试添加 .js / .json /.node
// 5.Module._cache 默认会判断是否存在缓存
// 6.new Module 创建模块(对象)   id ,exports
// 7.把模块缓存起来，方便下次使用
// ______________________ 根据文件名（绝对路径） 创建一个模块
// 8.tryModuleLoad 尝试加载模块   module.load
// 9.module.paths 第三方模块查找的路径
// 10.获取当前模块的扩展名  根据扩展名调用对应的方法Module._extensions 策略模式
// 11.获取文件的内容
// 12.调用module._compile 方法
// 13.将用户的内容 包裹到一个函数中  (function (exports, require, module, __filename, __dirname) {})

// 最终返回的是module.exports  用户会给这个module.exports进行赋值
const path = require('path');
const fs = require('fs');
const vm = require('vm');



function Module(id) {
    this.id = id;
    this.exports = {};
}
Module.wrap = function (script) {
    let arr = [
        ' (function (exports, require, module, __filename, __dirname) {',
        script,
        '})'
    ]
    return arr.join('')
}
Module._extensions = {
    '.js': function(module) {
        let content = fs.readFileSync(module.id,'utf8');
        let fnStr = Module.wrap(content);
        let fn = vm.runInThisContext(fnStr);
        let exports = module.exports;
        let require = myRequire;
        let __filename = module.id;
        let __dirname = path.dirname(module.id);
        // 这里的this 就是exports对象
        fn.call(exports,exports,require,module,__filename,__dirname);
        // 用户会给module.exports 赋值
    },
    '.json': function(module) {
        let content = fs.readFileSync(module.id,'utf8'); // 读取出来的是字符串
        module.exports = JSON.parse(content);
    }
}
Module._resolveFilename = function(filepath) {
    // 根据当前路径实现解析
    let filePath = path.resolve(__dirname, filepath);
    // 判断当前面文件是否存在
    let exists = fs.existsSync(filePath);
    if (exists) return filePath; // 如果存在直接返回路径

    // 尝试添加后缀
    let keys = Object.keys(Module._extensions);
    for (let i = 0; i < keys.length; i++) {
        let currentPath = filePath + keys[i];
        if (fs.existsSync(currentPath)) { // 尝试添加后缀查找
            return currentPath;
        }
    }
    throw new Error('模块不存在')
}
Module.prototype.load = function(filename) {
    // 获取文件的后缀来进行加载
    let extname = path.extname(filename);
    Module._extensions[extname](this); // 根据对应的后缀名进行加载
}
Module.cache = {};
Module._load = function(filepath) {
    // 将路径转化成绝对路径
    let filename = Module._resolveFilename(filepath);

    // 获取路径后不要立即创建模块，先看一眼能否找到以前加载过的模块
    let cacheModule = Module.cache[filename];
    console.log(cacheModule);
    if(cacheModule){
        return cacheModule.exports; // 直接返回上一次require的结果
    }

    // 保证每个模块的唯一性，需要通过唯一路径进行查找
    let module = new Module(filename); // id,exports对应的就是当前模块的结果
    Module.cache[filename] = module
    module.load(filename);
    return module.exports;
}
function myRequire(filepath) {
    // 根据路径加载这个模块
    return Module._load(filepath)
}
// myRequire = require

setInterval(() => {
    let r = myRequire('./a.js');
    console.log(r,'----');
}, 1000);


// 1.require语法是同步的，fs.readFileSync
// 2.最终require语法返回的是module.exports
// 3.模块的exports 和 module.exports 引用的是同一个变量
// 4.模块是"动态"加载每次require都会获取最新的导出的结果,可以将require写到条件中
// 5.更改exports的引用 不会导致module.exports变化
// 6.循环引用，一般不会出现，如果出现只能加载部分数据


