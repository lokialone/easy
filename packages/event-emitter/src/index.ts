interface funcContentInterface {
    func: Function, once: boolean
}
class EventEmitter {
    private events: {
        [k: string] : funcContentInterface[]
    };
    constructor() {
        this.events = {};
    }
    //
    private _on(eventName: string, func: Function, once: boolean) {
        let funcContent :funcContentInterface = {func, once}
        if (this.events[eventName]){
            this.events[eventName].push(funcContent)
        }  else {
            this.events[eventName] = [funcContent];
        }
    };
    on(eventName: string, func: Function) {
        this._on(eventName, func, false)
    }

    once(eventName: string, func: Function) {
        this._on(eventName, func, true);
    }

    emit(eventName: string) {
        let funcs = this.events[eventName];
        funcs.forEach(funcContent => {
            funcContent.func();
        });
        //去除执行一次的函数
        let notOnce = funcs.filter(funcContent => !funcContent.once);
        this.events[eventName] = notOnce;
    }

    off(eventName: string, func?: Function) {
        if (func) {
            let funcs = this.events[eventName];
            let remain = funcs.filter(funcContent => funcContent.func !== func);
            this.events[eventName] = remain;
        } else {
            this.events[eventName] = [];
        }
        
    }
}

export default EventEmitter;