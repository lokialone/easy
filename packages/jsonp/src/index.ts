import JSONP from './jsonp';

JSONP({
    url: 'http://localhost:3000/jsonp',
    // data: {  
    //     key1: 'key1'  
    // },
    callback: (json) => {
        console.log(json);
    }
})