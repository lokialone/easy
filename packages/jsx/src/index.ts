const nodeTypes = {
    element: 'element',
    value: 'value',
    props: 'props',
}
interface INode {
    type: string,
    length: number,
    value?: string,
    props?: {
        [k: string]: any
    },
    name?: string,
    children?: INode[],
}

export const parseElement = (str: string) => {
    let match = str.match(/<(\w+)/);
    // 判断为 value node or element node
    if (!match) {
        str = str.split('<')[0];
        return parseValue(str);
    }

    const node: INode = {
        type: nodeTypes.element,
        props: parseProps(''),
        children: [],
        length: 0,
        name: ''
    }

    node.name = match[1];
    let length = match.index ?  match.index + match[0].length : match[0].length;
    str = str.slice(length);
    node.length += length;

 
    match = str.match(/>/);

    if (!match) return node;
    // ??
    // node.props = parseProps(str.slice(0, match.index), values)
    node.props = parseProps(str.slice(0, match.index))
    length = node.props.length
    str = str.slice(length);
    node.length += length;


    // 判断是否是自闭合tag
    match = str.match(/^ *\/ *>/);
    if (match) {
        if (match.index === undefined) match.index = 0;
        node.length += match.index + match[0].length;
        return node
    }
    // get children
    match = str.match(/>/)
    if (!match) return node
    if (match.index === undefined) match.index = 0;
    length = match.index + 1
    str = str.slice(length);
    node.length += length;

    let child = parseElement(str);
    while(child.type === nodeTypes.element || child.value) {
        length = child.length;
        str = str.slice(length);
        node.length += length;
        if (!node.children) {
            node.children = [];
        }
        node.children.push(child);
        child = parseElement(str);
    }


   
    return node;
}

const parseProps = (str: string) => {
    let match;
    let length = '';

    const node: INode = {
        type: nodeTypes.props,
        length: 0
    }
    const matchNextProp = (str: string) => {
        return str.match(/ *\w+="(?:.*[^\\]")?/) || str.match(/ *\w+/);
    }
    match = matchNextProp(str);
    while(match) {
        const propStr = match[0];
        let [key, ...value] = propStr.split('=')
        node.length += propStr.length;
        key = key.trim();
        let valueString = value.join('=');
        node.props = {};
        node.props[key]= valueString? valueString.slice(1, -1) : true;
        if (match.index === undefined) match.index = 0;
        str = str.slice(0, match.index) + str.slice(match.index + propStr.length);
        match = matchNextProp(str);
    }

    return node;
}

const parseValue = (str: string) => {
    return {
        type: nodeTypes.value,
        length: str.length,
        value: str.trim()
    }
}

// const jsx = (splits, ...values) => {

// }

