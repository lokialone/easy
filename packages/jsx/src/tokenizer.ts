const TOKENS = ['<', '>', '=', '/'];
const WHITESPACE = /\s+/;
const isSingleQuote = /\'/;
const isDoubleQuote = /\"/;
// const NUMBER = /^[0-9]+$/;
// const STRING = /[0-9]

export default function tokenizer(html: string) {
    let length = html.length;
    let cursor = 0;
    let tokens = [];
    let tmpContent = '';
    function savePreviousContent() {
        if (tmpContent) {
            tokens.push(tmpContent);
            tmpContent = '';
        }
    }

    function readSingleQuteString() {
        let char = html.charAt(++cursor);
        let res = '';
        while(!isSingleQuote.test(char) && cursor < length) {
            res += char;
            char = html.charAt(++cursor);
        }
        tokens.push('\'' + res + '\'');
    }

    function readDoubleQuteString() {
        let char = html.charAt(++cursor);
        let res = '';
        while(!isDoubleQuote.test(char) && cursor < length) {
            res += char;
            char = html.charAt(++cursor);
        }
        tokens.push('"' + res + '"');
    }

    function readExpressions() {
        tokens.push('{');
        let char = html.charAt(++cursor);
        let res = '';
        while(char !== '}' && cursor < length) {
            res += char;
            char = html.charAt(++cursor);
        }
        tokens.push(res);
        tokens.push('}');
    }

    while (cursor < length) {
        let char = html.charAt(cursor);
        if (char === '{') {
            savePreviousContent();
            readExpressions();
            cursor++;
            continue;
        }
        if (isSingleQuote.test(char)) {
            savePreviousContent();
            readSingleQuteString();
            cursor++;
            continue;
            
        }
        if (isDoubleQuote.test(char)) {
            savePreviousContent();
            readDoubleQuteString();
            cursor++;
            continue;
        }

        if (TOKENS.includes(char)) {
            savePreviousContent();
            cursor++;
            tokens.push(char);
            continue;
        }
        if (WHITESPACE.test(char)) {
            savePreviousContent();
            cursor++;
            continue;
       }
       tmpContent += char;
       cursor++;
    } 
    return tokens;
}