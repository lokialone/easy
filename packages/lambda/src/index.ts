function partial<T>(func: Function, arg:T) {
    return (...args: T[]) => {
        return func.apply(null, [arg, ...args]);
    }
}

function sum (a: number, b: number):number {
    return a + b;
}

const a =  partial(sum, 2);
console.log(a(7));
