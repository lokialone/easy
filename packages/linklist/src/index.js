class Node {
    constructor(data, next) {
        this.data = data;
        this.next = next;
    }
}
class Linklist {
    constructor() {
        this.head = null;
        this.size  = 0;
    }

    add(index, ele) {
        if (arguments.length === 1) {
            ele = index;
            if (!this.size){
                this.head = new Node(ele, null);;
            } else {
                let prev = this.getNodeByIndex(this.size - 1);
                prev.next = new Node(ele, null);
            }
            this.size++  
        } else {
            if (index > this.size) {
                throw new Error('index > size');
            }
            
            if (this.size === 0 || index === 0){
                this.head = new Node(ele, this.head);
            } else {
                let prev = this.getNodeByIndex(index - 1);
                prev.next = new Node(ele, prev.next);
            }
            this.size++
        }
    }
    remove() {

    }
    getNodeByIndex(index) {
        if (index === 0) return this.head;
        let node = this.head;
        for(let i=0; i < index;i++) {
            node = node.next;
        }
        return node;
    }
}

const link = new Linklist();
link.add(1);
link.add(2);
link.add(3);
 
// console.dir(link,{depth: 10});


function reverse(link) {
    // console.log(link.data, link.data
    let newHead = null;
    let current = link.head;
    let next;
    while(current) {
        next = current.next;
        current.next = newHead;
        newHead = current;
        current = next;
    }
    link.head = newHead;
    return link;  
}

function reverse2(head) {
    // console.log(link.data, link.data
    if (!head || !head.next) return head;
    let current = reverse2(head.next);
    head.next.next = head;
    head.next = null;
    console.log('current', current);
    return current;
}

console.dir(reverse2(link.head), { depth: null });

const l1 = new Linklist();
l1.add(1);
l1.add(3);

const l2 = new Linklist();
l2.add(2);
l2.add(4);
function mergeLinkList(l1, l2) {
    if (!l1) return l2;
    if (!l2) return l1;
    if (l1.data > l2.data) {
        l2.next = mergeLinkList(l1, l2.next);
        return l2;
    } else {
        l1.next = mergeLinkList(l1.next, l2);
        return l1;
    }
}
// console.dir(mergeLinkList(l1.head, l2.head), {depth:null});