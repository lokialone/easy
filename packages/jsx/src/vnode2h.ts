import tokenizer from './tokenizer'

const WHITESPACE = /\s+/;
type IAttribute = {
    type: 'var' | 'string' | 'null',
    value: string | null
}
interface IAttributes {
    [k: string]: IAttribute
}
interface IVnode {
    name?: string;
    attributes?: IAttributes,
    children?: (string | IVnode)[]
}

export function readAttributes(tokens:string[], cursor: number) {
    let char = tokens[cursor];
    let attrs: IAttributes= {};
    let isValue = false; //根据等号前后来判断
    let attr:IAttribute = {
        type: 'null',
        value: null
    };
    let selfEnd = false;
    while(char !== '>' && cursor < tokens.length) {
        if (WHITESPACE.test(char)) {
            char = tokens[++cursor];
            continue;
        };

        if (char === '=') {
            isValue = true;
            char = tokens[++cursor];
            continue;
        }

        if (char === '{') {
            attr.type = 'var';
            char = tokens[++cursor];
            continue;
        }
        if (char === '}') {
            char = tokens[++cursor];
            continue;
        }
        if (char === '/' && tokens[++cursor] === '>') {
            selfEnd = true;
            break;
           
        } else {
            cursor--;
        }

        if(!isValue) {
            attr = attrs[char] = {
                type: 'null',
                value: null, 
            };
        } else {
            if (attr.type === 'null') attr.type = 'string';
            attr.value = char;
            isValue = false;
        }
        char = tokens[++cursor];
    }
    return  {
        _cursor: cursor + 1,
        attrs,
        selfEnd
    }
}

// const isTagStart = () =>

function parseChildren(tokens: string[], cursor: number = 0) {
    let result: (string| IVnode)[] = [];
    console.log(cursor, tokens[cursor]);
    while(cursor < tokens.length) {
        if (tokens[cursor] === '<' && tokens[cursor + 1] !== '/') {
            let children: IVnode= {};
            children.name = tokens[cursor + 1];
            const { _cursor, attrs, selfEnd } = readAttributes(tokens, cursor + 2);
            cursor = _cursor;
            children.attributes = attrs;
            result.push(children);
            if (!selfEnd) {
                children.children = parseChildren(tokens, cursor);
            }
            return result;
        }
        if (tokens[cursor] === '<' && tokens[cursor + 1] === '/') {
            cursor += 2;
            while(tokens[cursor] !== '>' && cursor < tokens.length) {
                cursor++;
            }
            cursor++;
            continue;
        }

        result.push(tokens[cursor]);
        cursor ++;

    }
    return result;
}

export function toVnode(tokens: string[], result: any) {
    let cursor = 0;
    if (tokens[cursor] === '<' && tokens[cursor + 1] !== '/') {
        result.name = tokens[cursor + 1];
        const { _cursor, attrs, selfEnd } = readAttributes(tokens, cursor + 2);
        cursor = _cursor;
        result.attributes = attrs;
        result.children = [];
        if (!selfEnd) {
            result.children = parseChildren(tokens, cursor);
        }
    } else {
        console.error('请输入root标签');
    }

    
    return result;
    // console.log(result);
}
let data = `
<div onClick={()=>this.click}>{hello} world</div>
`;
// console.log(tokenizer(data))
console.log(toVnode(tokenizer(data),{}));
// export function parse(str: string) {
//     return 
// }

// export default parse;