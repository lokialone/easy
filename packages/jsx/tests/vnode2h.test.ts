import { toVnode } from '../src/vnode2h';
import tokenizer from '../src/tokenizer';

test('toVnode div onClick', () => {
    let data = `
        <div onClick={()=>this.click}>{hello} world</div>
    `;
    // expect(toVnode(tokenizer(data), {})).toMatchSnapshot();
    // console.log(toVnode(tokenizer(data), {}));
});

test('toVnode children', () => {
    let data = `
        <div><span>hello</span></div>
    `;
    // expect(toVnode(tokenizer(data), {})).toMatchSnapshot();
    console.log(toVnode(tokenizer(data), {}));
});
