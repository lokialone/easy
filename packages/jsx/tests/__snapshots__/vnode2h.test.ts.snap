// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toVnode children 1`] = `
Object {
  "attributes": Object {},
  "children": Array [
    Object {
      "attributes": Object {},
      "children": Array [
        "hello",
      ],
      "name": "span",
    },
  ],
  "name": "div",
}
`;
